### PR TITLE
Use ax_cxx_compile_stdcxx.m4 and move to gnu++14

### PR DESCRIPTION
**Pull request by namtsui**
Use ax_cxx_compile_stdcxx.m4 instead of ax_cxx_compile_stdcxx_11.m4. Before, it defaulted to -std=c++11 (conformance mode). Now, use extended mode (e.g., gnu++14) and newer c++14 standard like cmake.

### CODE REVIEW

**Comment by namtsui** [comment]
With the new m4 scripts in place, the configure will create a Makefile with -std=c++11 since it uses the older  ax_cxx_compile_stdcxx_11.m4 instead of ax_cxx_compile_stdcxx.m4.

Moving to -std=gnu++14 makes it more in line with cmake.

My concerns are:

- Is it a good idea to impose extended mode with [ext] (gnu++14) instead of conformance mode [noext] (c++14)?
- Is extended mode required by libtorrent-rasterbar?

related:
https://github.com/arvidn/libtorrent/issues/4999
https://github.com/arvidn/libtorrent/issues/5024

**Comment by arvidn** [comment]
I'm not a fan of enabling the gnu extensions. libtorrent does not use any of them (deliberately), so I prefer `c++14` over `gnu++14`

**Comment by arvidn** [comment]
as for the issue of building qBittorrent with C++14 and linking against libtorrent built with C++11. This is caused by an ABI issue I introduced (a bit recklessly), and there's now a patch in 1.2.9 where a client can define `TORRENT_CXX11_ABI` to still be able to link against a C++11-build of libtorrent.

It would probably be best to build libtorrent with C++14 as well, if available.

**Comment by hhoffstaette** [comment]
Hi, it seems Gentoo just stumbled over this problem when trying to build with std=c++14 (see [here](https://bugs.gentoo.org/739654)) and I'd like to request some advice. Is there some quick intermediate fix we can do to make the Python binding observe the imposed C++ ABI until this PR is merged properly?

**Comment by namtsui** [comment]
@hhoffstaette Similarly for my proposed OpenBSD port update I had to decide between:
- sed c++11 c++14 for every Makefile, including bindings/python/compile_cmd and bindings/python/Makefile. I saw that you had noted the compile_cmd file in your Gentoo bug report.
- introducing autotools with hardcoded 14 in configure.ac
```
-AX_CXX_COMPILE_STDCXX_11([noext], [mandatory])
+AX_CXX_COMPILE_STDCXX(14, [noext], [mandatory])
```
Either temporary fix works. I ended up going with introducing autotools. I would like to clean up this draft a bit since it isn't quite where I want. Ideally configure would be able to specify the C++ dialect.

**Comment by arvidn** [comment]
this is still marked as a draft, but it sounds like it has important changes that ought to land for the 1.2.11. Is that right?

**Comment by eli-schwartz** [comment]
The archlinux package maintainer has just imported this patch as-is to fix our distro package. Would be good to have it merged for the next release...

**Comment by namtsui** [comment]
@arvidn Thanks for the ping. This is ready for review now. I vastly simplified it by keeping the m4 macro default. There is now an inline C++ program to test for __cplusplus.

The m4 macro already contains a test, but no harm in testing it twice:
```
// If the compiler admits that it is not ready for C++11, why torture it?
// Hopefully, this will speed up the test.
```

There are some limitations in the implementation though.

- User can't specify --with-cxx-standard=20. This is probably fine because ax_cxx_compile_stdcxx.m4 only supports 11, 14 and 17. In this case, specifying anything >17 or <11 results in 11, like point 4 above.
- Always prefers c++ over gnu++ even if the compiler default is gnu++. This is fine because no gnu extensions are needed.

Responding inline to the flowchart above:
-     If the compiler's default is equal to or greater than c++11/gnu++11, don't add anything, respect the compiler's default.

handled by the else case: `dnl nothing specified so use compiler default`. Still fails this point if given 20a.

-     If the compiler's default is less than c++11/gnu++11, use -std=c++11/-std=gnu++11
handled by the last `AX_CXX_COMPILE_STDCXX_11([noext], [optional])`

-     If the user manually overrides the C++ standard value with -DCMAKE_CXX_STANDARD at configure-time, and this value is greater than or equal to c++11/gnu++11, use it, regardless of compiler default.

handled by first 3 `AX_CXX_COMPILE_STDCXX(17` `AX_CXX_COMPILE_STDCXX(14` `AX_CXX_COMPILE_STDCXX_11`. Still fails this point if given 20a.

-     If the user manually overrides the value with -DCMAKE_CXX_STANDARD at configure-time, and this value is less than c++11/gnu++11, use -std=c++11/-std=gnu++11 instead, regardless of compiler default.

handled by two elif clauses around `dnl force 11 if user specified out of bounds, regardless of compiler default`

Overall, I think this is a good tradeoff between lowering maintenance for libtorrent-rasterbar (by not deviating too far from gnu m4 macros) and emulating cmake to an extent.

**Comment by namtsui** [comment]
@eli-schwartz Can you let archlinux know that the pull request was updated? I am concerned that every time I commit it will break the archlinux package due to not matching checksum.

https://github.com/archlinux/svntogit-packages/blob/packages/libtorrent-rasterbar/trunk/PKGBUILD

**Comment by namtsui** [comment]
C++ standard tester:
```
#include <iostream>

using std::cout;
using std::endl;

int main() {
	switch (__cplusplus) {
		case 201103L:
			cout << 11 << endl;
			break;
		case 201402L:
			cout << 14 << endl;
			break;
		case 201703L:
			cout << 17 << endl;
			break;
		default:
			cout << 1 << endl;
	}
}
```

**Comment by eli-schwartz** [comment]
@felixonmars ^^

**Comment by namtsui** [comment]
noticed a bug where it doesn't respect default compiler when no flag is specified because it falls into:
```
elif test "${ARG_CXX_STANDARD}" -lt "11" ; then
  dnl force 11 if user specified out of bounds, regardless of compiler default
  ARG_CXX_STANDARD="11"
  AX_CXX_COMPILE_STDCXX_11([noext], [mandatory])
elif test "${ARG_CXX_STANDARD}" -gt "17" ; then
  dnl force 11 if user specified out of bounds, regardless of compiler default
  ARG_CXX_STANDARD="11"
  AX_CXX_COMPILE_STDCXX_11([noext], [mandatory])
```

**Comment by namtsui** [comment]
some unit tests:

no --with-cxx-standard specified:
cxx standard:         14
c++ -std=c++14

--with-cxx-standard=14:
cxx standard:         14
c++ -std=c++14

--with-cxx-standard=11:
cxx standard:         11
c++ -std=c++11

17:
cxx standard:         17
c++ -std=c++17

2a:
cxx standard:         14
c++ -std=c++14

98:
cxx standard:         11
c++ -std=c++11

03:
cxx standard:         11
c++ -std=c++11

20:
cxx standard:         11
c++ -std=c++11

**Comment by emmercm** [comment]
I think this PR is still relevant, I don't believe v1.2.12 fixed it.

**Comment by namtsui** [comment]
It is still ready for review. I can answer any questions.

**Comment by arvidn** [comment]
@namtsui oh, I missed that. sorry

**Review by arvidn — COMMENTED** [review]
works for me on ubuntu

**Review by arvidn — APPROVED** [review]

**Comment by arvidn** [comment]
@namtsui Do you want to change the commits or should I just squash them all?
Do you think it makes sense to mention this in the Changelog?

**Comment by namtsui** [comment]
> @namtsui Do you want to change the commits or should I just squash them all?
> Do you think it makes sense to mention this in the Changelog?

Hopefully I edited history correctly. Yeah it could be a good idea to let configure users know of this new --with-cxx-standard=14 flag.

**Pull request merged**
